pallets/parachain-staking/src/lib.rs

```diff
@@ -834,34 +838,163 @@ pub mod pallet {
-		pub fn leave_candidates(origin: OriginFor<T>) -> DispatchResultWithPostInfo {
+		#[pallet::weight(<T as pallet::Config>::WeightInfo::init_leave_candidates(T::MaxCollatorCandidates::get(), T::MaxCollatorCandidates::get() * T::MaxDelegatorsPerCollator::get()))]
+		pub fn init_leave_candidates(origin: OriginFor<T>) -> DispatchResultWithPostInfo {
 			let collator = ensure_signed(origin)?;
 			let mut state = <CollatorState<T>>::get(&collator).ok_or(Error::<T>::CandidateNotFound)?;
 			ensure!(!state.is_leaving(), Error::<T>::AlreadyLeaving);
-			let mut exits = <ExitQueue<T>>::get();
+
 			let now = <Round<T>>::get().current;
 			let when = now.saturating_add(T::ExitQueueDelay::get());
-			ensure!(
-				exits.insert(Stake {
-					owner: collator.clone(),
-					amount: when
-				}),
-				Error::<T>::AlreadyLeaving
-			);
 			state.leave_candidates(when);
 			let mut candidates = <CandidatePool<T>>::get();
 			if candidates.remove_by(|stake| stake.owner.cmp(&collator)).is_some() {
 				<CandidatePool<T>>::put(candidates);
 			}
-			<ExitQueue<T>>::put(exits);
 			<CollatorState<T>>::insert(&collator, state);
 
 			// update candidates for next round
 			let (num_collators, num_delegators, _, _) = Self::select_top_candidates();
 
 			Self::deposit_event(Event::CollatorScheduledExit(now, collator, when));
-			Ok(Some(<T as pallet::Config>::WeightInfo::leave_candidates(
+			Ok(Some(<T as pallet::Config>::WeightInfo::init_leave_candidates(
+				num_collators,
+				num_delegators,
+			))
+			.into())
+		}
+
+		/// Execute the network exit of a candidate who requested to leave at
+		/// least `ExitQueueDelay` rounds ago. Prepares unstaking of the
+		/// candidates and their delegators stake which can be withdrawn via
+		/// `withdraw_unstaked` after `StakeDuration`.
+		///
+		/// Requires the candidate to previously have called
+		/// `init_leave_candidates`.
+		///
+		/// The exit request can be reversed by calling
+		/// `cancel_leave_candidates`.
+		///
+		/// Emits `CollatorLeft`.
+		///
+		/// # <weight>
+		/// Weight: O(D) where D is the number of delegators of the collator
+		/// candidate bounded by `MaxDelegatorsPerCollator`
+		/// - Reads: CollatorState, Round, D * DelegatorState, D
+		///   * BlockNumber, D * Unstaking
+		/// - Writes: D * Unstaking, D * DelegatorState, Total
+		/// - Kills: CollatorState, DelegatorState
+		/// # </weight>
+		// TODO: Add benchmark
+		#[pallet::weight(100_000_000_000)]
+		pub fn execute_leave_candidates(origin: OriginFor<T>) -> DispatchResultWithPostInfo {
+			let collator = ensure_signed(origin)?;
+			let state = <CollatorState<T>>::get(&collator).ok_or(Error::<T>::CandidateNotFound)?;
+			ensure!(state.is_leaving(), Error::<T>::NotLeaving);
+			ensure!(state.can_exit(<Round<T>>::get().current), Error::<T>::CannotLeaveYet);
+
+			// iterate over delegators
+			let _num_delegators = state.delegators.len() as u32;
```

Comment:
for weight? or is it unused?

Reply:
Yes, this will be used for post weight reduction.